### ISSUE TITLE
[RFC]: Fix + Standardize DiT Config/Init Patterns

### ISSUE THREAD

**Issue by alex-jw-brooks** [vllm-project/vllm-omni#1456]
### Motivation.

This will fix existing bad behaviors in DiT initialization around default values from Diffusers and hopefully add a sustainable pattern for well-typed DiT configs - this will also handle merging parsed config values with default values in a way that keeps the model code as clear as possible.

### Proposed Change

This is mostly an internal change for preventing bugs in config parsing / DiT initialization and making the code more clear, so there will be no changes to accuracy, performance, or APIs.

## Summary of Current Problem
There are currently some issues with how configs are being handled, which partially stems from the way in which Diffusers manages its configs compared to Transformers (i.e., through the use of mapping to classes via mixins rather than having easily inspectable `PretrainedConfig` classes per architecture).

In the current pipeline for Diffusion models, most components are created `.from_pretrained`, but the Transformer component is not initialized in a way that is consistent. There are generally 3 ways that we initialize it:

1. By passing the omni diffusion config and pulling things off thge DiT config from in the class
2. By exploding the DiT into kwargs derived from the DiT config
3. By relying on defaults in the `__init__` of the transformer class

The last of which is particularly problematic, because it causes classes to be initialized incorrectly. The defaults are copied from the corresponding Diffusers class, but since these tend to be overridden from the subconfigs they're loaded, they actually may not be correct.

## Example - LongCat
The current LongCat config management illustrates this issue well;

- In the current Image pipeline, we pass the Diffusion config [here](https://github.com/vllm-project/vllm-omni/blob/82e1bf2804784f1dfa6977e106df19937344675e/vllm_omni/diffusion/models/longcat_image/pipeline_longcat_image.py#L238)

```python
self.transformer = LongCatImageTransformer2DModel(od_config=od_config)
```

While the intializer takes many of these values [here](https://github.com/vllm-project/vllm-omni/blob/82e1bf2804784f1dfa6977e106df19937344675e/vllm_omni/diffusion/models/longcat_image/longcat_image_transformer.py#L517):
```
    def __init__(
        self,
        od_config: OmniDiffusionConfig,
        patch_size: int = 1,
        in_channels: int = 64,
        num_layers: int = 19,
        num_single_layers: int = 38,
        attention_head_dim: int = 128,
        num_attention_heads: int = 24,
        joint_attention_dim: int = 3584,
        pooled_projection_dim: int = 3584,
        axes_dims_rope: list[int] = [16, 56, 56],
    ):
```

whose kwargs are used to set the attributes on the class, not the passsed transformer config. These values are understandably taken from the default of the corresponding Diffusers class [here](https://github.com/huggingface/diffusers/blob/aac94befce95f1842511cf13f43b2b60fa27b966/src/diffusers/models/transformers/transformer_longcat_image.py#L413).

However, AFAIK they are not the right values for *any* LongCat model; I think they are copied from Flux as the reference architecture. E.g,. the Transformers config for [LongCatImage](https://huggingface.co/meituan-longcat/LongCat-Image/blob/main/transformer/config.json).
```json
{
    "_class_name": "LongCatImageTransformer2DModel",
    "_diffusers_version": "0.30.0.dev0",
    "attention_head_dim": 128,
    "guidance_embeds": false,
    "in_channels": 64,
    "joint_attention_dim": 3584,
    "num_attention_heads": 24,
    "num_layers": 10,
    "num_single_layers": 20,
    "patch_size": 1,
    "pooled_projection_dim": 3584
}
```

As such, the pipeline ends up incorrectly configured, and in this case much heavier than it should be, because it initializes extra transformer blocks that it doesn't have weights for. I noticed this while working on TeaCache support for LongCat since I was seeing nans and unexpected OOMS while calibrating coefficients through the transformers block. This is the approach currently used for diffusion pipelines.

## Proposed Solution
I think this can be fixed and extended to be well-patterned in two PRs.

1. Short term fix - standardize all transformers to take only take the `OmniDiffusionConfig` as input to initialization, and pull the expected attributes off of the config.

This will fix the models that are currently not initialized correctly, e.g., due to different configs for the same arch, but may not be very pretty since we do need to ensure we fall back to defaults where needed, e.g., for `axes_dims_rope` in `LongCat` - I think a reasonable temporary fix is to have a default config dict for each supported DiT class that is clobbered by values in the `OmniDiffusionConfig`'s `tf_model_config` with warning logs if the `_class_name` is not as expected in each model.

2. Longer term pattern - I think this should be handled similarly to the way in which vLLM handles custom config types, since it is similar.

## Example - Kimi-VL's Config in vLLM
`Kimi-VL` has its own custom config defined ([example](https://huggingface.co/moonshotai/Kimi-VL-A3B-Thinking/blob/main/configuration_kimi_vl.py)), and it's also using a copied `MoonViTConfig` in its own config file.

To handle this, vLLM has its own version of the Kimi-VL config ([ref](https://github.com/vllm-project/vllm/blob/542ca66357e7128d43f67b2013c7adfce77829d1/vllm/transformers_utils/configs/kimi_vl.py#L11)), which wraps around its own `MoonViTConfig`, and is cleanly mapped to based on config type [here](https://github.com/vllm-project/vllm/blob/542ca66357e7128d43f67b2013c7adfce77829d1/vllm/transformers_utils/configs/__init__.py#L46). This allows for better type hints and explicit checks against the nonstandard config type ([example](https://github.com/vllm-project/vllm/blob/542ca66357e7128d43f67b2013c7adfce77829d1/vllm/model_executor/models/kimi_vl.py#L313)).

I think that we should take a similar pattern here, and have a registry which maps the expected Diffuser `_class_name`s to concrete DiT config types in vLLM Omni, e.g., under `vllm_omni.diffusion.dit_configs`, and ensure that the `tf_model_config` is the proper subtype; this will let us encapsulate the default values, write generic utils to merge with defaults prior to interacting with the corresponding model pipeline class, and allow us to have a clear pattern for how to init the ViT that will still be useful if the existing pipeline loading mechanism is changed, e.g., to use the vLLM weight loader. It will also make it more clear why things are going wrong if an unexpected `_class_name` is used.

Opening this RFC for community discussion - I think both of these are fast to implement and am planning to open PRs for both based on feedback, but will open the short term fix for bad initialization behaviors in existing models first so that potential architectural changes can be decoupled from the bugfixes.

### Feedback Period.

One week - will open the first fix for bad behaviors in the next day or so, and will likely link a draft at the longer term pattern to this in the next week as well, since they should both be fast to implement

### CC List.

@DarkLight1337 @Isotr0py @ywang96 @sfeng33 @NickLucche @hsliuustc0106 @Gaohan123 @tzhouam @ZJY0516 in case anyone has thoughts

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
Let me ask @hmellor about this, see if it's possible to fix this from HF side.

**Comment by alex-jw-brooks** [comment]
Cool thanks @DarkLight1337 - I spent some time on this today and opened a draft PR for a few models [here](https://github.com/vllm-project/vllm-omni/pull/1466). CC'ed you both on it in case it's easier to discuss concretely

**Comment by hmellor** [comment]
This is really valuable feedback and definitely something that seems would be better fixed on the Diffusers side.

I've forwarded this RFC to the Diffusers team, I'll let you know their response if they don't do it directly here.

**Comment by sayakpaul** [comment]
Hey folks,

Sayak here from the Diffusers team 👋 

Currently, on `diffusers`, we can do:

```py
from diffusers import AutoModel

ckpt_id = "black-forest-labs/FLUX.1-dev"
config = AutoModel.load_config(ckpt_id, subfolder="transformer")
model = AutoModel.from_config(config)
```

Of course, one can change this `config` object to pass custom values.

Does this solve this inconsistency in the caller sites?

We can work on doing better defaults, but note that most of these are taken directly from the original checkpoints. Some inconsistencies can arise due to the fact that `Flux2Transformer2DModel` can be used for both Flux2 and Flux2 Klein (although the difference here is just in the number of layers).

**Comment by alex-jw-brooks** [comment]
Hey @sayakpaul, thanks for the quick response! I think there are two issues this RFC intends to address:

1. Standardizing the pattern for how DiTs in vLLM Omni are initialized and removing existing bugs that exist because it's not well-patterned yet
2. Making model development for DiTs in vLLM Omn well-patterned & cleaner

The second one is the main intention with adding well-defined configs classes, because underneath, we are currently doing pretty much the same thing, i.e.,
```python
            ...
            tf_config_dict = get_hf_file_to_dict("transformer/config.json", od_config.model) # `json.load`s the downloaded file
            od_config.tf_model_config = TransformerConfig.from_dict(tf_config_dict) # takes the dict and puts it in a generic wrapper
```

The main advantages for having the concrete types with defaults set is the following:
1. Clarity & type hints for contributors writing the DiTs
2. It decouples the vLLM Omni DiT implementations from the Diffusers implementations, which will make managing versions easier
   - E.g., if the Diffusers DiT implementation modified config key(s), it could be cleanly handled to remap correctly to Diffusers. However, since we are just loading a `dict` in this case and initializing our own DiT with it, we would need to handle this correctly on our side atm, which may not be obvious without clear typing
3. All values are well-defined; currently we have some things like this in DiTs
```python
        self.dual_attention_layers = (
            model_config.dual_attention_layers if hasattr(model_config, "dual_attention_layers") else ()
        )
        self.qk_norm = model_config.qk_norm if hasattr(model_config, "qk_norm") else ""
```
This is not ideal, because in the refactor example above, we would not be able to easily catch something like `dual_attention_layers` is not in the config, so we should use `()` vs `dual_attention_layers` was renamed to something else, but since it isn't there, we use the default value of `()` anyway 🙂 This is another reason why something like [this](https://github.com/alex-jw-brooks/vllm-omni/blob/a12a9bac1b051c15ce1b55ad499924e3fe800e9a/vllm_omni/diffusion/config/sd3.py) is IMO much more maintainable, because we can more easily add validation to check for potential key misalignment and not have to rely on `hasattr`/`getattr` for things that may be excluded in the raw config file

**Comment by sayakpaul** [comment]
Thanks for the continued and elaborate discussions! PFB my thoughts.

> The main advantages for having the concrete types with defaults set is the following:

We can work with the `vllm-omni` team to improve typing so that it's a better developer experience. Would you be willing to pair with me on this? For defaults, I believe we already have sensible defaults on most of our implementations (at least the modern ones from the last ~12 months). 

One exception is the case where a modeling class like `Flux2Transformer2DModel` is used for both Flux2 and Flux2 Klein. This is how the original codebase also implements it and IMO, it makes sense because they both differ just in terms of the number of layers.

> E.g., if the Diffusers DiT implementation modified config key(s), it could be cleanly handled to remap correctly to Diffusers. However, since we are just loading a dict in this case and initializing our own DiT with it, we would need to handle this correctly on our side atm, which may not be obvious without clear typing

I am not too sure about it. It would be helpful if you could provide a representative example of this. If better typing on `diffusers` would help alleviate this issue, we're of course, up for it.

> All values are well-defined; currently we have some things like this in DiTs

I agree this is not ideal and we're happy to standardize on this on `diffusers`. I believe this has largely been the case for the models added in the last ~12 months in `diffusers`. If there's bad residue, we should fix that. Could you share some examples where this is happening?

In general, init arguments are rarely renamed in `diffusers` unless we have an extremely strong reason to do so. We're happy to keep you folks in the loop if and when those cases arise in the future. This way, we can brainstorm the solutions together. WDYT?

**Comment by hsliuustc0106** [comment]
> We can work with the `vllm-omni` team to improve typing so that it's a better developer experience. Would you be willing to pair with me on this? For defaults, I believe we already have sensible defaults on most of our implementations (at least the modern ones from the last ~12 months).

cc @ZJY0516 @wtomin

**Comment by alex-jw-brooks** [comment]
For sure @sayakpaul, thanks! LongCat Image is a good example for both unhappy defaults and not setting all of the kwargs in diffuseres; it has wrong defaults for `num_single_layers` / `num_layers `, and the [config](https://huggingface.co/meituan-longcat/LongCat-Image/blob/main/transformer/config.json) doesn't set `axes_dim_ropes` to pass [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/transformers/transformer_longcat_image.py#L423).

The layers thing is an example of a currently existing copy paste from our side that will fixed by [this](https://github.com/vllm-project/vllm-omni/pull/1485), but `axes_dim_ropes` isn't in the resulting dict when you load the config for example.

```python
ckpt_id = "meituan-longcat/LongCat-Image"
config = AutoModel.load_config(ckpt_id, subfolder="transformer")
```

```
{'_class_name': 'LongCatImageTransformer2DModel', '_diffusers_version': '0.30.0.dev0', 'attention_head_dim': 128, 'guidance_embeds': False, 'in_channels': 64, 'joint_attention_dim': 3584, 'num_attention_heads': 24, 'num_layers': 10, 'num_single_layers': 20, 'patch_size': 1, 'pooled_projection_dim': 3584}
```

Maybe an easy place to start on the Diffusers side that we could potentially iterate on would be a utility that
1. Gets the the config as above ^
2. Uses the retrieved `config["_class_name"]` to determine the DiT class being initialized
3. `inspects` the retrieved class for default kwargs
4. returns the `{**config, **model_defaults}` to apply the overrides and missing keys

This does unfortunately keep things as a raw `dict`, but at least would ensure incorrect defaults are overridden in the same way, and that all expected kwargs to init are in the `dict` if we have want to know what the DiT actually would be using if the same config were used in Diffusers. I am also happy to open a PR in Diffusers to discuss, if it would help!

**Comment by sayakpaul** [comment]
This is all very reasonable and detailed feedback. I want to start by acknowledging that. Thank you! Please keep doing this :)

Sometimes, during the integration process, the authors themselves lead them and we might have to prioritize speed for supporting the release (which I believe was the case for LongCat). This likely explains the reason why you saw it for LongCat but not for other modern models. But this doesn't excuse the friction that gets introduced in the developer experience (+ the inconsistency).

> LongCat Image is a good example for both unhappy defaults and not setting all of the kwargs in diffuseres; it has wrong defaults for num_single_layers / num_layers , and the [config](https://huggingface.co/meituan-longcat/LongCat-Image/blob/main/transformer/config.json) doesn't set axes_dim_ropes to pass [here](https://github.com/huggingface/diffusers/blob/main/src/diffusers/models/transformers/transformer_longcat_image.py#L423).

We can definitely open a PR to fix that config and work with the author to get that merged.

> This does unfortunately keep things as a raw dict, but at least would ensure incorrect defaults are overridden in the same way, and that all expected kwargs to init are in the dict if we have want to know what the DiT actually would be using if the same config were used in Diffusers. I am also happy to open a PR in Diffusers to discuss, if it would help!

Of course, we will be quick to review and work with you to get that prioritized.

**Comment by DN6** [comment]
Hi @alex-jw-brooks. I think there are a couple of options to get the required config. 

If you load the model `from_config` on meta device, you could access the config from the model instance without the overhead of creating weights.  

```python
import torch
from diffusers import AutoModel

repo = "meituan-longcat/LongCat-Image"
subfolder = "transformer"

config = AutoModel.load_config(repo, subfolder=subfolder)

with torch.device("meta"):
    model = AutoModel.from_config(config)
print(f"model.config:")
for k, v in dict(model.config).items():
    if not k.startswith("_"):
        print(f"  {k}: {v}")

```

Would print
```shell
model.config:
  patch_size: 1
  in_channels: 64
  num_layers: 10
  num_single_layers: 20
  attention_head_dim: 128
  num_attention_heads: 24
  joint_attention_dim: 3584
  pooled_projection_dim: 3584
  axes_dims_rope: [16, 56, 56]
  guidance_embeds: False
```

The keys and value types would correspond with exactly what the init expects. 

We could explore having a utility method that returns a typed dataclass config for the model. e.g. 
```py
from diffusers import AutoModel

config_dict = AutoModel.load_config(
    "meituan-longcat/LongCat-Image",
    subfolder="transformer",
)
# import DiT based on _class_name
typed_config = LongCatImageTransformer2DModel._create_dataclass_from_config(config_dict)
for f in dataclasses.fields(typed_config):
     print(f"{f.name}: {f.type}")
```
But I think inspecting the key names and types in the FrozenDict that we currently return with the `config` property would be just as effective as a dataclass since all the init args have to be JSON serializable anyway.
